### PR TITLE
ability to build and load custom segment in realtime node

### PR DESCRIPTION
**Pull request by kaijianding**
It's a follow up of #2965 #3901

1. introduce sinkFactory to use custom sink to handle events, like not using onHeapIncrementalIndex to hold events, but still using storageAdapter interface when query
2. introduce customIndexMerger to build custom segment when persist and merge
3. can load custom segment when bootstrapSinksFromDisk and after persist success

### CODE REVIEW

**Comment by leventov** [comment]
@kaijianding you are doing a lot of PRs, could you also review other people's PRs? And then become a committer?

**Comment by kaijianding** [comment]
Sure, will review other's PRs and glad to hear that I'm close to being a committer @leventov

**Comment by leventov** [comment]
@kaijianding I'm not sure about closeness and it's not for me to decide because I'm not a PMC member, but making reviews is essential because it's a painful aspect in the project that not enough reviews are done. When you are making a lot of PRs you demand reviews but not "contribute them back".

**Comment by KurtYoung** [comment]
I will try to review this. Sorry for not having much time to do the review.

**Comment by gianm** [comment]
cc @cheddar @pjain1 who may be able to help review as well due to previous work in the area.

**Comment by pjain1** [comment]
I can give it a try next week

**Comment by kaijianding** [comment]
annoy unrelated Travis failure: The job exceeded the maximum time limit for jobs, and has been terminated.
will retry after #4468 and #4466

File: extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/KafkaTuningConfig.java [hunk 1]
```diff
@@ -80,6 +86,8 @@ public KafkaTuningConfig(
     this.resetOffsetAutomatically = resetOffsetAutomatically == null
                                     ? DEFAULT_RESET_OFFSET_AUTOMATICALLY
                                     : resetOffsetAutomatically;
+    this.sinkFactory = sinkFactory == null ? defaults.getSinkFactory() : sinkFactory;
+    this.customIndexMerger = customIndexMerger;
```
Comment (KurtYoung) [inline extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/KafkaTuningConfig.java]: Do we need to handle null here?

File: extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/KafkaTuningConfig.java [hunk 2]
```diff
@@ -273,6 +310,8 @@ public String toString()
            ", reportParseExceptions=" + reportParseExceptions +
            ", handoffConditionTimeout=" + handoffConditionTimeout +
            ", resetOffsetAutomatically=" + resetOffsetAutomatically +
+           ", sinkFactory=" + sinkFactory +
+           ", customIndexMerger=" + customIndexMerger +
```
Comment (KurtYoung) [inline extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/KafkaTuningConfig.java]: IndexMerger doesn't provide proper toString() method

Comment (kaijianding) [inline extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/KafkaTuningConfig.java]: IndexMerger is interface now, the implement of this interface should implement toString

File: extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/KafkaTuningConfig.java (hunk 1 above)
Comment (kaijianding) [inline extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/KafkaTuningConfig.java]: the caller will handle the null customIndexMerger. like in IndexTask: 
```
        ingestionSchema.getTuningConfig().getCustomIndexMerger() != null
        ? ingestionSchema.getTuningConfig().getCustomIndexMerger() : toolbox.getIndexMergerV9()
```
customIndexMerger is a customized IndexMerger which can be null if user don't need it. 
If we set a default IndexMerger implement to customIndexMerger, then I think customIndexMerger should be renamed to `IndexMerger` which should never be null according to the literal meaning

File: indexing-service/src/main/java/io/druid/indexing/common/task/MergeTask.java [hunk 3]
```diff
@@ -65,30 +66,33 @@ public MergeTask(
...
+        .transform(
+            new Function<Map.Entry<DataSegment, File>, QueryableIndex>()
             {
               @Override
-              public QueryableIndex apply(@Nullable File input)
+              public QueryableIndex apply(Map.Entry<DataSegment, File> input)
               {
                 try {
-                  return toolbox.getIndexIO().loadIndex(input);
+                  return toolbox.getIndexIO().loadSegment(input.getKey(), input.getValue()).asQueryableIndex();
```
Comment (KurtYoung) [inline indexing-service/src/main/java/io/druid/indexing/common/task/MergeTask.java]: What is the differences between loadSegment().asQueryableIndex() and just loadIndex()?
I noticed there still exists plenty usage of loadIndex(), should they also be changed?

File: indexing-service/src/test/java/io/druid/indexing/common/task/SameIntervalMergeTaskTest.java [hunk 4]
```diff
@@ -68,14 +69,13 @@
   private final CountDownLatch publishCountDown = new CountDownLatch(1);
   private final IndexSpec indexSpec;
   private final ObjectMapper jsonMapper;
-  private IndexIO indexIO;
 
   public SameIntervalMergeTaskTest()
   {
+    EmittingLogger.registerEmitter(new NoopServiceEmitter());
```
Comment (KurtYoung) [inline indexing-service/src/test/java/io/druid/indexing/common/task/SameIntervalMergeTaskTest.java]: Do we need this?

File: java-util/src/main/java/io/druid/java/util/common/io/smoosh/Metadata.java [hunk 5]
```diff
@@ -21,7 +21,7 @@
 
 /**
 */
-class Metadata
+public class Metadata
```
Comment (KurtYoung) [inline java-util/src/main/java/io/druid/java/util/common/io/smoosh/Metadata.java]: Do we need this?

File: java-util/src/main/java/io/druid/java/util/common/io/smoosh/SmooshedFileMapper.java [hunk 6]
```diff
@@ -97,7 +97,7 @@ public static SmooshedFileMapper load(File baseDir) throws IOException
   private final Map<String, Metadata> internalFiles;
   private final List<MappedByteBuffer> buffersList = Lists.newArrayList();
 
-  private SmooshedFileMapper(
+  public SmooshedFileMapper(
```
Comment (KurtYoung) [inline java-util/src/main/java/io/druid/java/util/common/io/smoosh/SmooshedFileMapper.java]: Do we need this?

File: processing/src/main/java/io/druid/segment/IndexIO.java [hunk 7]
```diff
@@ -118,6 +127,24 @@ public void validateTwoSegments(File dir1, File dir2) throws IOException
     }
   }
 
+  public Segment loadSegment(DataSegment dataSegment, File segmentDir) throws IOException
+  {
+    return loadSegment(dataSegment.getIdentifier(), segmentDir);
+  }
+
+  public Segment loadSegment(String identifier, File segmentDir) throws IOException
```
Comment (KurtYoung) [inline processing/src/main/java/io/druid/segment/IndexIO.java]: Does this related to your proposed change?

File: processing/src/main/java/io/druid/segment/StringDimensionMergerV9.java [hunk 8]
```diff
@@ -120,17 +122,19 @@ public void writeMergedValueMetadata(List<IndexableAdapter> adapters) throws IOE
     long dimStartTime = System.currentTimeMillis();
 
     this.adapters = adapters;
+    this.dimValuesList = toDimValueList(adapters);
```
Comment (KurtYoung) [inline processing/src/main/java/io/druid/segment/StringDimensionMergerV9.java]: Looks like this is some kind of performance improvement?

File: indexing-service/src/test/java/io/druid/indexing/common/task/SameIntervalMergeTaskTest.java (hunk 4 above)
Comment (kaijianding) [inline indexing-service/src/test/java/io/druid/indexing/common/task/SameIntervalMergeTaskTest.java]: yes, otherwise the test will fail.
```
  public AlertBuilder makeAlert(Throwable t, String message, Object... objects)
  {
    if (emitter == null) {
      final String errorMessage = String.format(
          "Emitter not initialized!  Cannot alert.  Please make sure to call %s.registerEmitter()", this.getClass()
      );
```
the emitter is registered in EmittingLogger.registerEmitter

File: java-util/src/main/java/io/druid/java/util/common/io/smoosh/Metadata.java (hunk 5 above)
Comment (kaijianding) [inline java-util/src/main/java/io/druid/java/util/common/io/smoosh/Metadata.java]: yes, the customIndexMerger may need to use Metadata, like in my customIndexMerger implement.

File: java-util/src/main/java/io/druid/java/util/common/io/smoosh/SmooshedFileMapper.java (hunk 6 above)
Comment (kaijianding) [inline java-util/src/main/java/io/druid/java/util/common/io/smoosh/SmooshedFileMapper.java]: my implement want to sub class SmooshedFileMapper, so need it can be extended.

File: processing/src/main/java/io/druid/segment/IndexIO.java (hunk 7 above)
Comment (kaijianding) [inline processing/src/main/java/io/druid/segment/IndexIO.java]: yes, this is used to load persisted segment in realtime node

File: processing/src/main/java/io/druid/segment/StringDimensionMergerV9.java (hunk 8 above)
Comment (kaijianding) [inline processing/src/main/java/io/druid/segment/StringDimensionMergerV9.java]: this is some cleanup to StringDimensionMergerV9.java, in this part, only dimValuesList is needed, but more info is required(the adapters) than needed.

File: indexing-service/src/main/java/io/druid/indexing/common/task/MergeTask.java (hunk 3 above)
Comment (kaijianding) [inline indexing-service/src/main/java/io/druid/indexing/common/task/MergeTask.java]: every place want to load index from file should call loadSegment instead of loadIndex. inside `loadSegment`, it will use the customized segmentizerFactory to map the file to index.
Most usages of loadIndex() are in tests and benchmarks, so I don't modify them.
I will double check the rest usages to make sure every place calling loadIndex is replaced by loadSegment()

**Comment by kaijianding** [comment]
The purpose of this PR is to provide the ability to build and load customized segment. SinkFactory and customizeIndexMerger are used to build customized segment, and the indexIo.loadSegment() is to load the customized segment  So I think the 3 modifications are better in one PR. @KurtYoung

File: indexing-service/src/main/java/io/druid/indexing/common/task/MergeTask.java (hunk 3 above)
Comment (KurtYoung) [inline indexing-service/src/main/java/io/druid/indexing/common/task/MergeTask.java]: If `loadSegment` is the only correct way to load a segment, then i would suggest you just delete the `loadIndex' method.

File: indexing-service/src/test/java/io/druid/indexing/common/task/SameIntervalMergeTaskTest.java (hunk 4 above)
Comment (KurtYoung) [inline indexing-service/src/test/java/io/druid/indexing/common/task/SameIntervalMergeTaskTest.java]: Can you double check this? I commented this line and the test passed. BTW, why this test will not fail before you added this, and why the test case would enter the logic about making any alerts?

File: java-util/src/main/java/io/druid/java/util/common/io/smoosh/Metadata.java (hunk 5 above)
Comment (KurtYoung) [inline java-util/src/main/java/io/druid/java/util/common/io/smoosh/Metadata.java]: Can we change this until we see a real customized index merger?

File: java-util/src/main/java/io/druid/java/util/common/io/smoosh/SmooshedFileMapper.java (hunk 6 above)
Comment (KurtYoung) [inline java-util/src/main/java/io/druid/java/util/common/io/smoosh/SmooshedFileMapper.java]: Same above

File: processing/src/main/java/io/druid/segment/StringDimensionMergerV9.java (hunk 8 above)
Comment (KurtYoung) [inline processing/src/main/java/io/druid/segment/StringDimensionMergerV9.java]: So this is unrelated to this PR, right? Can we separate this to another PR?

**Comment by KurtYoung** [comment]
Make sense except the third point. Overall it looks good to me. @leventov Do you have any design concerns?

File: indexing-service/src/main/java/io/druid/indexing/common/task/MergeTask.java (hunk 3 above)
Comment (kaijianding) [inline indexing-service/src/main/java/io/druid/indexing/common/task/MergeTask.java]: I rewrite the indexIO.loadIndex() to use segmentizerFactory to load index from file, then there is no need to modify all the place calling indexIO.loadIndex() now

File: indexing-service/src/test/java/io/druid/indexing/common/task/SameIntervalMergeTaskTest.java (hunk 4 above)
Comment (kaijianding) [inline indexing-service/src/test/java/io/druid/indexing/common/task/SameIntervalMergeTaskTest.java]: after rewrite indexIO.loadIndex(), there is no need to modify this anymore

File: java-util/src/main/java/io/druid/java/util/common/io/smoosh/Metadata.java (hunk 5 above)
Comment (kaijianding) [inline java-util/src/main/java/io/druid/java/util/common/io/smoosh/Metadata.java]: ok

File: processing/src/main/java/io/druid/segment/IndexIO.java (hunk 7 above)
Comment (kaijianding) [inline processing/src/main/java/io/druid/segment/IndexIO.java]: change it to loadIndex()

File: extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/KafkaTuningConfig.java [hunk 9]
```diff
@@ -44,6 +46,8 @@
   @Deprecated
   private final long handoffConditionTimeout;
   private final boolean resetOffsetAutomatically;
+  private final SinkFactory sinkFactory;
```
Comment (leventov) [inline extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/KafkaTuningConfig.java]: Please extract an umbrella abstraction like "SegmentStrategy" with createSinkFactory() and createIndexMerger() methods, and inject/serialize/deserialize only it, here and in other places.

File: extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/KafkaTuningConfig.java [hunk 10]
```diff
@@ -241,8 +269,15 @@ public boolean equals(Object o)
         : that.basePersistDirectory != null) {
       return false;
     }
-    return indexSpec != null ? indexSpec.equals(that.indexSpec) : that.indexSpec == null;
-
+    if (indexSpec != null ? !indexSpec.equals(that.indexSpec) : that.indexSpec != null) {
```
Comment (leventov) [inline extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/KafkaTuningConfig.java]: Refactor to use Objects.equals()

File: extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/KafkaTuningConfig.java [hunk 11]
```diff
@@ -257,6 +292,8 @@ public int hashCode()
     result = 31 * result + (reportParseExceptions ? 1 : 0);
     result = 31 * result + (int) (handoffConditionTimeout ^ (handoffConditionTimeout >>> 32));
     result = 31 * result + (resetOffsetAutomatically ? 1 : 0);
+    result = 31 * result + (sinkFactory != null ? sinkFactory.hashCode() : 0);
```
Comment (leventov) [inline extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/KafkaTuningConfig.java]: Refactor to use Objects.hash()

File: indexing-service/src/main/java/io/druid/indexing/common/index/YeOldePlumberSchool.java [hunk 12]
```diff
@@ -113,6 +114,10 @@ public Plumber findPlumber(
     // Set of spilled segments. Will be merged at the end.
     final Set<File> spilled = Sets.newHashSet();
 
+    // IndexMerger implementation.
```
Comment (leventov) [inline indexing-service/src/main/java/io/druid/indexing/common/index/YeOldePlumberSchool.java]: Not useful comment

File: indexing-service/src/main/java/io/druid/indexing/common/index/YeOldePlumberSchool.java [hunk 13]
```diff
@@ -113,6 +114,10 @@ public Plumber findPlumber(
...
+    final IndexMerger theIndexMerger = config.getCustomIndexMerger() != null
```
Comment (leventov) [inline indexing-service/src/main/java/io/druid/indexing/common/index/YeOldePlumberSchool.java]: Regarding null handling - doesn't Jackson handle it up front, when you specify `defaultImpl=...` on the `IndexMerger` interface (or SegmentStrategy for that matter, if you apply suggestion above)?

File: java-util/src/main/java/io/druid/java/util/common/io/smoosh/SmooshedFileMapper.java (hunk 6 above)
Comment (leventov) [inline java-util/src/main/java/io/druid/java/util/common/io/smoosh/SmooshedFileMapper.java]: @kaijianding `SmooshedFileMapper` is not a part of public api, and could be changed anytime so that your subclass will break. Worse, the behaviour may break, however without compilation errors, so you won't notice it. So if you want something similar to `SmooshedFileMapper` in your custom SegmentizerFactory, I strongly suggest to copy `SmooshedFileMapper` entirely into your codebase and tailor it for your needs.

File: processing/src/main/java/io/druid/segment/IndexIO.java [hunk 14]
```diff
@@ -91,12 +93,13 @@
   private static final SerializerUtils serializerUtils = new SerializerUtils();
 
   private final ObjectMapper mapper;
+  private final ColumnConfig columnConfig;
```
Comment (leventov) [inline processing/src/main/java/io/druid/segment/IndexIO.java]: The intention is the opposite, to move column config out of IndexIO: #1078, not to establish it as a part of IndexIO. And actually this PR is a good moment to move ColumnConfig at least one step down, to specific "SegmentStrategies" (those that need it).

File: processing/src/main/java/io/druid/segment/IndexIO.java [hunk 15]
```diff
@@ -192,6 +205,25 @@ public QueryableIndex loadIndex(File inDir) throws IOException
...
 
+  public SegmentizerFactory getSegmentizerFactory(File segmentDir) throws IOException
+  {
+    File factoryJson = new File(segmentDir, "factory.json");
+    final SegmentizerFactory factory;
+
+    if (factoryJson.exists()) {
+      factory = mapper.readValue(factoryJson, SegmentizerFactory.class);
+    } else {
+      factory = new MMappedQueryableSegmentizerFactory(this);
```
Comment (leventov) [inline processing/src/main/java/io/druid/segment/IndexIO.java]: Could be singleton or cached.

File: processing/src/main/java/io/druid/segment/IndexIO.java [hunk 16]
```diff
@@ -410,7 +442,7 @@ public MMappedIndex mapDir(File inDir) throws IOException
 
   }
 
-  static interface IndexLoader
+  public static interface IndexLoader
```
Comment (leventov) [inline processing/src/main/java/io/druid/segment/IndexIO.java]: Why public?

File: processing/src/main/java/io/druid/segment/IndexIO.java [hunk 17]
```diff
@@ -519,7 +551,7 @@ public QueryableIndex load(File inDir, ObjectMapper mapper) throws IOException
     }
   }
 
-  static class V9IndexLoader implements IndexLoader
+  public static class V9IndexLoader implements IndexLoader
```
Comment (leventov) [inline processing/src/main/java/io/druid/segment/IndexIO.java]: Why public?

File: processing/src/main/java/io/druid/segment/IndexIO.java [hunk 18]
```diff
@@ -181,6 +189,11 @@ public void validateTwoSegments(final IndexableAdapter adapter1, final Indexable
 
   public QueryableIndex loadIndex(File inDir) throws IOException
   {
+    return getSegmentizerFactory(inDir).loadIndex(inDir);
+  }
+
+  public QueryableIndex loadIndexDirectly(File inDir) throws IOException
```
Comment (leventov) [inline processing/src/main/java/io/druid/segment/IndexIO.java]: Readers and users will never know the difference between loadIndex() and loadIndexDirectly(). Maybe move loadIndexDirectly() to MMappedQueryableSegmentizerFactory. Along with some other methods.

File: processing/src/main/java/io/druid/segment/IndexMerger.java [hunk 19]
```diff
@@ -58,6 +60,10 @@
 import java.util.Set;
 import java.util.stream.Collectors;
 
+@JsonTypeInfo(use = JsonTypeInfo.Id.NAME, property = "type", defaultImpl = IndexMergerV9.class)
+@JsonSubTypes(value = {
+    @JsonSubTypes.Type(name = "v9", value = IndexMergerV9.class)
+})
 @ImplementedBy(IndexMergerV9.class)
```
Comment (leventov) [inline processing/src/main/java/io/druid/segment/IndexMerger.java]: `@ImplementedBy` is likely not needed since Json defaultImpl is added.

File: processing/src/main/java/io/druid/segment/StringDimensionMergerV9.java [hunk 20]
```diff
@@ -196,6 +200,20 @@ public void writeMergedValueMetadata(List<IndexableAdapter> adapters) throws IOE
     setupEncodedValueWriter();
   }
 
+  protected List<Indexed<String>> toDimValuesList(List<IndexableAdapter> adapters) throws IOException
```
Comment (leventov) [inline processing/src/main/java/io/druid/segment/StringDimensionMergerV9.java]: As far as I can understand you changed this class in order to be able to extend it in your code. See comment to `SmooshedFileMapper`, don't do this. Copy class entirely into your codebase.

File: processing/src/test/java/io/druid/query/aggregation/AggregationTestHelper.java [hunk 21]
```diff
@@ -174,7 +174,7 @@ public static final AggregationTestHelper createSelectQueryAggregationTestHelper
         new InjectableValues.Std().addValue(
```
Comment (leventov) [inline processing/src/test/java/io/druid/query/aggregation/AggregationTestHelper.java]: Formatting
```java
new InjectableValues.Std()
  .addValue(SelectQueryConfig.class, new SelectQueryConfig(true))
  .addValue(IndexIO.class, TestHelper.getTestIndexIO())
```

File: java-util/src/main/java/io/druid/java/util/common/io/smoosh/SmooshedFileMapper.java (hunk 6 above)
Comment (kaijianding) [inline java-util/src/main/java/io/druid/java/util/common/io/smoosh/SmooshedFileMapper.java]: I want to use V9IndexLoader.load(SmooshedFileMapper smooshedFiles, File inDir, ObjectMapper mapper), this method requires SmooshedFileMapper.
It's fine SmooshedFileMapper can change behaviors at any time, I extend SmooshedFileMapper in my extension and will have unit test to ensure everything is working properly.

File: extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/KafkaTuningConfig.java (hunk 9 above)
Comment (kaijianding) [inline extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/KafkaTuningConfig.java]: Sure.

File: extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/KafkaTuningConfig.java (hunk 10 above)
Comment (kaijianding) [inline extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/KafkaTuningConfig.java]: it's generated by Intellij Idea, but sure, will modify it.

File: extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/KafkaTuningConfig.java (hunk 11 above)
Comment (kaijianding) [inline extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/KafkaTuningConfig.java]: sure

**Comment by gianm** [comment]
Just a drive by comment on the public api stuff. I believe none of the segment-type extension points are public apis at this point (see http://druid.io/docs/latest/development/modules.html) and that's because we're not yet entirely sure what custom segment extensions should look like and what support they might need from core druid.

IMO, if @kaijianding is ok with this then it's fine for his custom extensions to use internal apis. It's part of the learning process of what should and shouldn't be public for these kinds of extensions. At some point in the future, we might have a better idea of what should be public, and then we can make those apis public.

File: processing/src/main/java/io/druid/segment/IndexIO.java (hunk 18 above)
Comment (kaijianding) [inline processing/src/main/java/io/druid/segment/IndexIO.java]: ok, but many related code should be move to.

File: extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/KafkaTuningConfig.java (hunk 10 above)
Comment (leventov) [inline extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/KafkaTuningConfig.java]: You can generate the forms which that I suggested with intelliJ as well, need to choose another option in generation dialog. Once you choose it, it will be the default later.

**Comment by leventov** [comment]
@kaijianding I'm ok with just making (or keeping) classes and constructors public, but I'm not OK with adding properties and fields which are unused in the core codebase and protected methods to classes that have no subclasses in the core codebase. It increases complexity of the core for nothing. And it's not a step towards modular design, the ultimate goal of this PR as well.

File: processing/src/main/java/io/druid/segment/IndexMerger.java (hunk 19 above)
Comment (kaijianding) [inline processing/src/main/java/io/druid/segment/IndexMerger.java]: `@ImplementedBy` is guice thing, I'm not sure it can work properly after removing it if it is not a json case

File: processing/src/main/java/io/druid/segment/IndexIO.java (hunk 17 above)
Comment (kaijianding) [inline processing/src/main/java/io/druid/segment/IndexIO.java]: see comments below

File: processing/src/main/java/io/druid/segment/IndexIO.java (hunk 16 above)
Comment (kaijianding) [inline processing/src/main/java/io/druid/segment/IndexIO.java]: see comments below

File: processing/src/main/java/io/druid/segment/IndexIO.java (hunk 15 above)
Comment (kaijianding) [inline processing/src/main/java/io/druid/segment/IndexIO.java]: sure

File: processing/src/main/java/io/druid/segment/IndexIO.java (hunk 14 above)
Comment (kaijianding) [inline processing/src/main/java/io/druid/segment/IndexIO.java]: I'm not sure it's the correct point to move ColumnConfig out of IndexIO in this PR, because there are so many test code need be modified and they are irrelevant to this PR.
I think it's better do the refactor in another PR

**Comment by gianm** [comment]
Are there a lot of properties and fields in the patch that don't need to be protected or even exist at all? (Sorry, I haven't had a chance to read it yet). I'm hoping we can find a balance between keeping the core clean, and making the extension possible.

File: indexing-service/src/main/java/io/druid/indexing/common/index/YeOldePlumberSchool.java (hunk 13 above)
Comment (kaijianding) [inline indexing-service/src/main/java/io/druid/indexing/common/index/YeOldePlumberSchool.java]: usually the default implement is provided in the class context, like `indexMergerV9` here. Also the `TaskToolbox` has the default implement, like toolbox.getIndexMergerV9().
So I think it's better to regard the old code here to get the default implement.

**Comment by kaijianding** [comment]
Fine, will restore some changes and modify them when I provide an actually custom implement @leventov
